src/compute/src/render/reduce.rs

```diff
@@ -2125,6 +2189,63 @@ impl Columnation for Accum {
     type InnerRegion = CopyRegion<Self>;
 }
 
+/// Helper to merge key-value pairs into rows or split rows
+/// into key-value pairs provided a key arity.
+#[derive(Clone, Copy, Debug)]
+struct KeyValPairer {
+    key_arity: usize,
+}
+
+impl KeyValPairer {
+    fn new(key_arity: usize) -> Self {
+        Self { key_arity }
+    }
+
+    fn merge(&self, key: Row, val: Row) -> Row {
+        let binding = SharedRow::get();
+        let mut row_builder = binding.borrow_mut();
+        let mut row_packer = row_builder.packer();
+        row_packer.extend(&key);
+        row_packer.extend(&val);
+        row_builder.clone()
+    }
+
+    fn merge_val_u64<'a, I>(&self, key: I, val: u64) -> Row
+    where
+        I: Iterator<Item = Datum<'a>>,
+    {
+        let binding = SharedRow::get();
+        let mut row_builder = binding.borrow_mut();
+        let mut row_packer = row_builder.packer();
+        row_packer.push(Datum::UInt64(val));
+        row_packer.extend(key);
+        row_builder.clone()
+    }
```

Comment:
You might be able to combine this into one function that takes two iterators that yield datums. Not sure, but worth a try.